models/spectral_clustering.py

```diff
@@ -0,0 +1,280 @@
+		smax = np.iinfo(np.int16).max
+		f2s = lambda signal, max=np.float32(smax): torch.mul(signal, max).to(torch.int16)
+		s2f = lambda signal, max=np.float32(smax): torch.div(signal, max, dtype = torch.float32)
+
+		if signal.dtype == torch.float32 and self.vad.required_type == 'int16':
+			signal = f2s(signal)
+		if signal.dtype == torch.int16 and self.vad.required_type == 'float32':
+			signal = s2f(signal)
+
+		if self.vad.required_wrapper == np.array:
+			signal = signal.cpu().numpy()
+			silence_mask = torch.as_tensor(self.vad.detect(signal)[0], dtype = torch.bool, device = self.device)
+		else:
+			silence_mask = self.vad.detect(signal)[0]
+		return silence_mask
+
+	def convert_mask_to_emb_space(self, mask):
+		mask = mask.to(torch.float32)[None, None, :]
+		for kernel_size, stride, padding in zip(self.embed_model.kernel_sizes, self.embed_model.strides, self.embed_model.paddings):
+			if stride == 1:
+				continue
+			mask = F.avg_pool1d(mask, kernel_size, stride, padding)
+		return mask.squeeze() > (1 - self.vad_sensitivity)
+
+	def get_speaker_mask(self, signal: shapes.BT, sample_rate: int):
+		# extract features
+		with torch.no_grad():
+			silence: shapes.T = self.get_silence_mask(signal)
+
+			features: shapes.BCt = self.embed_model(signal.to(self.device), sample_rate)
+			features_mask: shapes.t = self.convert_mask_to_emb_space(silence)
+
+			# compute affinity
+			emb = features[0, :, features_mask].t()
+			K = self._wang_affinity(emb, row_threshold = 0.5)
+
+			# spectral clustering
+			W = (K > 0.5).float()
+			L = normalized_symmetric_laplacian(W)
+			eigvals, eigvecs = L.symeig(eigenvectors = True)
+			num_eigvecs = 9
+			eigvecs = eigvecs[:, 1: (1 + num_eigvecs)]
+			eigvecs_max = eigvecs.max()
+			fiedler_vector = torch.zeros(features_mask.shape[-1], dtype = torch.float32, device = self.device)
+			fiedler_vector[features_mask] = eigvecs[:, 0] / eigvecs_max
+
+			# interpolate to T space
+			fiedler_vector = torch.nn.functional.interpolate(fiedler_vector[None, None, :], signal.shape[-1]).squeeze()
+			speaker1 = (fiedler_vector < 0) & ~silence
+			speaker2 = (fiedler_vector > 0) & ~silence
+			silence = silence | ~(speaker1 | speaker2)
+
+			speaker_mask = torch.stack([silence, speaker1, speaker2])
+		return speaker_mask
+
+	def _wang_affinity(self, E, row_threshold = 0.50, shrinking = 0.01):
+		# Speaker Diarization with LSTM, Wang et al, https://arxiv.org/abs/1710.10468
+		# https://github.com/wq2012/SpectralCluster/blob/master/spectralcluster/spectral_clusterer.py
+		A = cosine_kernel(E)
+
+		# crop diagonalw
+		A.fill_diagonal_(0)
+		A.diagonal().copy_(A.max(dim=1).values)
+
+		# row-wise shrinking
+		A = torch.where(A > row_threshold * A.max(dim=1, keepdim=True).values, A, shrinking * A)
+
+		# symmetrization
+		A = torch.max(A, A.t())
+
+		# diffusion
+		A = A @ A
+
+		# row-wise normalization
+		A = A / A.max(dim=1, keepdim=True).values
+		A = torch.max(A, A.t())
+		return A
+
+
+class SincTDNN(nn.Module):
```

Comment:
Is SincTDNN model duplicated in `models/__init__.py` and `models/spectral_clustering.py`?